djvm/src/main/kotlin/net/corda/djvm/source/SourceClassLoader.kt

```diff
@@ -21,43 +22,84 @@ import java.nio.file.Path
 import java.nio.file.Paths
 import kotlin.streams.toList
 
+/**
+ * Base interface for API and user sources.
+ */
+interface Source : Closeable {
+    val loadResource: (String) -> URL?
+}
+
+/**
+ * This is the contract that our source of Java APIs must satisfy.
+ */
+interface ApiSource : Source
+
+/**
+ * This is the contract that the source of the user's own code must satisfy.
+ * It will almost certainly be a [ClassLoader] of some description.
+ */
+interface UserSource : Source {
```

Comment:
I dont really understand this interface. Also isn't there an opportunity for malicious behavior if loadClass returns a class that is different from the class actually represented in bytecode inside the URLs? How do you check for consistency?

Reply:
My goal here is to "abstract away" the usage of `URLClassLoader`, given that `SourceClassLoader` _must_ still be a `ClassLoader` for the sake of ASM. So the point of the `UserSource` interface is to provide a facade that may _or may not_ have a `URLClassLoader` behind it. The SGX `MemoryClassLoader` can be extended also to satisfy this interface without too much effort.

Yes, this does make the DJVM harder to configure correctly, but that's not its threat model.

Reply:
Fair enough on the assumption that the `loadClass` behaviour cannot be "malicious". 
However, I dont understand why the `URL` concept needs to be part of this abstraction. I.e.: why can't the user simply provides a stream of bytes containing the classes he wants to transform and let the DJVM do the rest? 
Note additionally that SGX enclave will not be directly a client of this interface, i.e.: we wont run the DJVM directly in an SGX enclave.

Reply:
I've kept the URLs only to preserve information, i.e. the `SourceClassLoader` still knows where all of its sources come from. (And also because I'm expecting the sources _still_ to be some kind of `ClassLoader`, even in Corda.)

Reply:
My point was that i would imagine would be easier for the client code to just provide the bytecode as a stream of byte, while the DJVM internall takes care of building a classloader on top of that. But i guess it could be considered more a "nice to have" rather than requirement.